File: tech_interview_preparation/data_structures/stack/Stack.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.insert(0, item)

    def pop(self):
        return self.items.pop(0)

    def is_empty(self):
        return self.size() == 0

    def peek(self):
        if not self.is_empty():
            return self.items[0]

    def size(self):
        return len(self.items)


stack = Stack()

print(stack.is_empty())

stack.push(1)
print(stack.items)
stack.push(2)
print(stack.items)
stack.push(3)
print(stack.items)
stack.push(4)
print(stack.items)
stack.push(5)
print(stack.items)

print(stack.is_empty())
print(stack.peek())

print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())

print(stack.is_empty())

print(stack.pop())

print(stack.is_empty())
print(stack.peek())

File: tech_interview_preparation/data_structures/stack/stack.js
// push
const push = (stack, item) => [item, ...stack];

// pop
const pop = (stack) => stack.slice(1);

// peek
const peek = (stack) => stack[0];

// isEmpty
const isEmpty = (stack) => !stack.length;

// size
const size = (stack) => stack.length;

// testing
let stack = [];

console.log(isEmpty(stack));

stack = push(stack, 1);
console.log(stack);
stack = push(stack, 2);
console.log(stack);
stack = push(stack, 3);
console.log(stack);
stack = push(stack, 4);
console.log(stack);
stack = push(stack, 5);
console.log(stack);

console.log(peek(stack));
console.log(size(stack));
console.log(isEmpty(stack));

stack = pop(stack);
stack = pop(stack);
stack = pop(stack);
stack = pop(stack);

console.log(isEmpty(stack));
console.log(peek(stack));
stack = pop(stack);
console.log(isEmpty(stack));
console.log(size(stack));
